src/DS/metrics_extraction.py

```diff
@@ -0,0 +1,300 @@
+from loguru import logger
+from pathlib import Path
+
+from utils.git_utils import get_working_directory_or_git_root
+
+BATCH_SIZE = 100  # Number of files to process before saving to disk
+
+# Configure logger
+logger.add("metrics_extraction_log", rotation="500 MB", level="INFO")
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description="Analyse Python test files and extract metrics."
+    )
+
+    parser.add_argument(
+        "--input",
+        type=str,
+        default=Path(
+            get_working_directory_or_git_root() / "data" / "guessed_languages_rows.csv"
+        ),
+        help="Path to the input CSV file.",
+    )
+
+    parser.add_argument(
+        "--output",
+        type=str,
+        default=Path(
+            get_working_directory_or_git_root()
+            / "data"
+            / "test_metrics_df_guesslang.csv"
+        ),
+        help="Path to the output CSV file.",
+    )
+
+    return parser.parse_args()
+
+
+def read_and_parse_file(file_path):
+    """
+    Reads and parses a Python file into an AST.
+
+    Args:
+        file_path (str): The path to the Python file.
+
+    Returns:
+        tuple: A tuple containing the AST tree and the file content, or
+        None if there's an error.
+    """
+    logger.debug(f"Attempting to read and parse: {file_path}")
+    if not file_path.endswith(".py"):
+        logger.info(f"Skipping non-python file {file_path}")
+        return None
+
+    try:
+        with open(file_path, "r", encoding="utf-8") as file:
+            content = file.read()
+            tree = ast.parse(content, filename=file_path)
+            logger.debug(f"Successfully parsed file: {file_path}")
+            return tree, content
+    except SyntaxError as e:
+        logger.error(f"Syntax error in file {file_path}: {e}")
+    except UnicodeDecodeError as e:
+        logger.error(f"Unicode decode error in file {file_path}: {e}")
+    except FileNotFoundError as e:
+        logger.error(f"File not found: {file_path}: {e}")
+    except Exception as e:
+        logger.error(f"Unexpected error processing file {file_path}: {e}")
+
+    return None
+
+
+def analyse_test_file(file_path):
+    """
+    Analyses a Python test file to extract the number of test cases,
+    assertions, presence of setup and teardown methods, and the complexity
+    of the test file.
+
+    Args:
+        file_path (str): The path to the test file.
+
+    Returns:
+           dict: A dictionary containing the analysis results.
+    """
+    logger.info(f"Starting analysis of test file: {file_path}")
+
+    result = {
+        "num_test_cases": 0,
+        "num_assertions": 0,
+        "has_setup": False,
+        "has_teardown": False,
+        "complexity": 0,
+    }
+
+    parsed_file = read_and_parse_file(file_path)
+    if parsed_file is None:
+        logger.warning(f"Analysis skipped for file: {file_path}")
+        return result
+
+    tree, _ = parsed_file
+
+    num_test_cases = 0
```

Comment:
As per my previous comment, how about initialising these to `-1`?

BTW why did you decide to use a replica set of local parameters rather than updating the various parameters in the `result` data structure?

Reply:
It was more clear in my idea as they’re decoupled from the result structure. I have changed the code to  directly  update the dictionary. This approach helps reducing the number of code lines as well. Thank you.